src/psyclone/psyir/nodes/loop.py

```diff
@@ -578,3 +583,109 @@ def gen_code(self, parent):
             if not kernel.module_inline:
                 parent.add(UseGen(parent, name=kernel._module_name, only=True,
                                   funcnames=[kernel._name]))
+
+    @property
+    def is_outermost(self):
+        '''
+        :returns: True if the loop has no ancestors, otherwise False
+        :rtype: bool
+        '''
+        return self.ancestor(Loop) is None
+
+    @property
+    def is_perfectly_nested(self):
+        '''
+        Determine whether the Loop nest of descendents of a Loop (inclusive) is
+        'perfect', i.e., each level excecpt the deepest contains only the next
+        Loop.
+
+        Note that we ignore Nodes of certain types.
+
+        :returns: True if the Loop nest is perfect, otherwise False
+        :rtype: bool
+        '''
+        exclude = (
+            Literal,
+            Reference,
+            Loop,
+            IntrinsicCall,
+        )
+        subnest = self.walk(Loop)
+
+        def intersect(list1, list2):
+            '''
+            Compute the intersection of two lists.
+
+            We cannot use the inbuilt set interesection in Python because
+            Nodes are not hashable.
+
+            :param list1: the first list to intersect
+            :type list1: list
+            :param list2: the second list to intersect
+            :type list2: list
+            :returns: the intersection of the two lists
+            :rtype: list
+            '''
+            return [item for item in list1 if item in list2]
+
+        # Check whether the subnest is perfect by checking each level in turn
```

Comment:
This looks pretty complicated. I think you'd be better off using the `loop_body` property of `Loop` rather than doing a `walk`. Could you use/adapt the implementation at:
https://github.com/stfc/PSyclone/blob/8367de1fbce19bbd0fa64e2dc2ad2bfcd0ed6e82/src/psyclone/psyir/transformations/parallel_loop_trans.py#L271-L325
(and then replace/update that code to use this new method)?

Reply:
Thanks @arporter, what do you think of 97d1143ce?